lib/server/middleware.js

```diff
@@ -0,0 +1,507 @@
+            contentLength: req.headers['content-length'],
+            contentType: req.headers['content-type'],
+            url: req.url,
+            isCompleteMultipartUpload: isCompleteMultipartUpload,
+            isBulkDeleteRequest: isBulkDeleteRequest,
+            isXmlRequest: isXmlRequest,
+            isCorsRequest: isCorsRequest,
+            isStsRequest: isStsRequest
+        }, 'S3_DEBUG: PRE-MIDDLEWARE - Preserving raw body for' +
+        ' S3 signature verification, bulk delete,' +
+        ' multipart upload, CORS configuration, or STS operations');
+
+        var chunks = [];
+
+        // Set up immediate data capture
+        req.pause(); // Pause the stream to prevent data loss
+
+        req.on('data', function (chunk) {
+            chunks.push(chunk);
+        });
+
+        req.on('end', function () {
+            if (chunks.length > 0) {
+                req._rawBodyBuffer = Buffer.concat(chunks);
+                req._rawBodyString = req._rawBodyBuffer.toString('utf8');
+
+                // Calculate and store the original content-md5
+                var md5Hash = crypto.createHash('md5');
+                md5Hash.update(req._rawBodyBuffer);
+                req._originalContentMD5 = md5Hash.digest('base64');
+
+                req.log.debug({
+                    rawBodyLength: req._rawBodyBuffer.length,
+                    originalContentMD5: req._originalContentMD5,
+                    headerContentMD5: req.headers['content-md5'],
+                    md5Match: req._originalContentMD5 ===
+                        req.headers['content-md5'],
+                    bodyPreview: req._rawBodyString.substring(0, 200)
+                }, 'S3_DEBUG: PRE-MIDDLEWARE - Raw body preserved' +
+                              ' for signature verification');
+
+                // Set req.body to the parsed content for immediate
+                // availability
+                req.body = req._rawBodyString;
+            }
+
+            // Call next() here to ensure body is available
+            // before proceeding
+            next();
+        });
+
+        req.on('error', function (err) {
+            req.log.error(err, 'S3_DEBUG: PRE-MIDDLEWARE'+
+                          ' - Error while preserving raw body');
+            next(err);
+        });
+
+        req.resume(); // Resume the stream
+
+        // Don't call next() here - wait for the 'end' event
+        return;
+    }
+    next();
+}
+
+
+/**
+ * Pre-middleware for early S3 request detection.
+ * Detects S3 requests by checking for SigV4 auth headers or presigned URLs,
+ * parses S3 request details, and marks binary operations.
+ *
+ * @param {Object} req - Restify request object
+ * @param {Object} res - Restify response object
+ * @param {Function} next - Restify next callback
+ */
+function s3RequestDetectorEarly(req, res, next) {
+    // Check if this is a SigV4 request and mark it for later processing
+    var authHeader = req.headers.authorization ||
```

Comment:
`req.header('Authorization', '')` is an equivalent drop-in replacement for this statement, which will fall back to an empty string if an authorization header (in either lower or capital case) is not provided.

The keys in `req.headers` are always lowercased by Node.js HTTP server so `req.headers.Authorization` will always be `undefined` here.